### PR TITLE
session handler info

### PR DESCRIPTION
**Pull request by nithu0115**
Issue #, if available: #409 

Description of changes:
With this PR, we will inject session handler information/user-agent information while creating the session. 

Tested with debug logging:
```
Host: api.ecr.us-west-2.amazonaws.com
**User-Agent: aws-controller-k8s/0c76963-dirty (GitCommit/0c76963771053ea4310b5ffd0aaa399af0f88e87; BuildDate/2020-10-14T00:33:32+0000; Kind/Repository; CRDVersion/v1alpha1) aws-sdk-go/1.34.32 (go1.14.1; linux; amd64)
Content-Length: 72**
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -5,11 +5,8 @@ ARG service_alias
 ARG work_dir=/github.com/aws/aws-controllers-k8s
 WORKDIR $work_dir
 # For building Go Module required
-ENV GOPROXY=https://proxy.golang.org,direct
-ENV GO111MODULE=on
-ENV GOARCH=amd64
-ENV GOOS=linux
-ENV CGO_ENABLED=0
+ENV GOPROXY=direct
```
Comment (A-Hilaly) [inline Dockerfile:8]: I'm curious, did you encounter any issue caused by the Go proxy ?
We made [this change a month ago](https://github.com/aws/aws-controllers-k8s/pull/314) because `gomodules.xyz` was down for more than 24h and was causing all the service controllers build to fail

File: Dockerfile [hunk 2]
```diff
@@ -19,10 +16,14 @@ RUN  go mod download
 # Copy the go source
 COPY . $work_dir/
 # Build
-RUN  go build -a -o $work_dir/bin/controller $work_dir/services/$service_alias/cmd/controller/main.go
+RUN GIT_VERSION=$(git describe --tags --dirty --always) && \
+    GIT_COMMIT=$(git rev-parse HEAD) && \
+    BUILD_DATE=$(date +%Y-%m-%dT%H:%M:%S%z) && \
+    CGO_ENABLED=0 GOOS=linux GOARCH=amd64 GO111MODULE=on go build \
+    -ldflags="-X ${VERSION_PKG}.GitVersion=${GIT_VERSION} -X ${VERSION_PKG}.GitCommit=${GIT_COMMIT} -X ${VERSION_PKG}.BuildDate=${BUILD_DATE}" -a -o $work_dir/bin/controller $work_dir/services/$service_alias/cmd/controller/main.go
```
Comment (A-Hilaly) [inline Dockerfile:23]: nit: i'm fine with this change but i find it easier to read when the environment variables are declared separately

**Review by A-Hilaly — COMMENTED** [review]
LGTM overall, leaving few

File: Dockerfile (hunk 1 above)
Comment (nithu0115) [reply Dockerfile:8]: Ah, yes. I kept running into sink-hole issue while I running builds with VPN. I modified this during testing and forgot to revert it back. Thanks for catching :D

**Review by kumargauravsharma — COMMENTED** [review]
As an AWS Service controller would have corresponding binary as container image; the current PR would have same user agent string for all service controller built against commit level.

How about having AWS service name also as part of user agent? Something like: `aws-controller-k8s/elasticache-0c76963-dirty`

**Comment by nithu0115** [comment]
> As an AWS Service controller would have corresponding binary as container image; the current PR would have same user agent string for all service controller built against commit level.
> 
> How about having AWS service name also as part of user agent? Something like: `aws-controller-k8s/elasticache-0c76963-dirty`

Done, addressed this comment. 
```
User-Agent: aws-controller-k8s/ecr.services.k8s.aws-b216274-dirty (GitCommit/b2162744daa8d5cb631a535df492c8e12254887e; BuildDate/2020-10-14T18:44:02-0500; CRDKind/Repository; CRDVersion/v1alpha1) aws-sdk-go/1.34.32 (go1.15.1; darwin; amd64)
```

**Review by kumargauravsharma — APPROVED** [review]

**Review by A-Hilaly — APPROVED** [review]

File: Dockerfile (hunk 2 above)
Comment (jaypipes) [reply Dockerfile]: I'm actually not fine with this change... @nithu0115 is there a reason you changed this like so instead of adding another environment variable?

File: Dockerfile [hunk 3]
```diff
@@ -19,10 +16,17 @@ RUN  go mod download
...
 # Build
-RUN  go build -a -o $work_dir/bin/controller $work_dir/services/$service_alias/cmd/controller/main.go
+RUN GIT_VERSION=$(git describe --tags --dirty --always) && \
+    GIT_COMMIT=$(git rev-parse HEAD) && \
+    BUILD_DATE=$(date +%Y-%m-%dT%H:%M) && \
+    CGO_ENABLED=0 GOOS=linux GOARCH=amd64 GO111MODULE=on \
+    go build -ldflags="-X ${VERSION_PKG}.GitVersion=${GIT_VERSION} \
+    -X ${VERSION_PKG}.GitCommit=${GIT_COMMIT} \
+    -X ${VERSION_PKG}.BuildDate=${BUILD_DATE}" \
+    -a -o $work_dir/bin/controller $work_dir/services/$service_alias/cmd/controller/main.go
```
Comment (jaypipes) [inline Dockerfile:26]: Is the reason you changed the above from using ENV variables to using CLI variables because you needed to execute shell commands to populate things like GIT_COMMIT? If so, why move the static env vars like CGO_ENABLED and GOOS to the command line invocation?

File: go.mod [hunk 4]
```diff
@@ -9,11 +9,11 @@ require (
...
-	github.com/google/go-cmp v0.3.0
+	github.com/google/go-cmp v0.4.0
 	github.com/iancoleman/strcase v0.0.0-20191112232945-16388991a334
 	github.com/mitchellh/go-homedir v1.1.0
 	github.com/pkg/errors v0.9.1
-	github.com/prometheus/client_golang v1.0.0
+	github.com/prometheus/client_golang v1.7.1
 	github.com/spf13/cobra v1.0.0
 	github.com/spf13/pflag v1.0.5
 	github.com/stretchr/testify v1.5.1
```
Comment (jaypipes) [inline go.mod:19]: please undo the changes to `go.mod` and `go.sum`. They are not germane to this pull request's content. If we want to update these packages, let's do that in a separate commit that is specific to the new code brought in with those new versions.

File: pkg/runtime/reconciler.go [hunk 5]
```diff
@@ -15,7 +15,6 @@ package runtime
 
 import (
 	"context"
-
```
Comment (jaypipes) [inline pkg/runtime/reconciler.go:18]: please restore this newline.

**Review by jaypipes — CHANGES_REQUESTED** [review]
@nithu0115 biggest change I need you to make on this is to undo the changes to go.mod and go.sum please.

File: Dockerfile (hunk 3 above)
Comment (nithu0115) [reply Dockerfile:26]: I thought as these are just build an argument, so I went ahead and added them as CLI variables. I can revert this change and add all of these as env vars.
